include/deal.II/matrix_free/dof_info.h

```diff
@@ -63,6 +63,7 @@ namespace internal
      *
      * @author Katharina Kormann and Martin Kronbichler, 2010, 2011, 2018
      */
+    template <typename Number = double>
     struct DoFInfo
```

Comment:
@kronbichler Heads up!

This is the most intrusive change that I made in order to support a templated `AffineConstraints` (aka `ConstraintMatrix`). I templatifyed `struct DoFInfo`. This change went remarkably smooth. There is only one minor caveat...

Reply:
This change goes a bit against the design of that class because it was designed to use the pure integer part, so a template at this place is confusing. Wouldn't it have been enough to simply put the template on the single function that passes in the AffineConstraints? On the other hand, I believe we should put most of that index functionality into DoFHandler anyway, as discussed in #2000, so we might accept this as a temporary hack. (Further assuming that we really do #2000 for the next release).

Reply:
test results indicate that the testsuite isn't quite happy with this.

I will reroll the matrix free portion (without a template parameter) tomorrow and give this another try. Probably following your suggestion and only templatifying the function that passes in the constraints.

Reply:
In any case - thanks a lot for the feedback :-)  (This kind of code review helps a lot. And to my defense - after working on the porting for 15h straight my brain was only half functional. So I didn't spot the obvious solution.)